Added Azure Log Analytics scaler documentation (#244)
kedacore/keda-docs: content/docs/2.0/scalers/azure-log-analytics.md

+++
title = "Azure Log Analytics"
layout = "scaler"
availability = "v2.0+"
maintainer = "Community"
description = "Scale applications based on Azure Log Analytics query result"
go_file = "azure_log_analytics_scaler"
+++

### Trigger Specification

> This specification describes the `azure-log-analytics` trigger for Azure Log Analytics query result (All auth parameters in metadata is not real, generated randomly).

Reviewer: ```suggestion
This specification describes the `azure-log-analytics` trigger for Azure Log Analytics query result.
```

```yaml
triggers:
  - type: azure-log-analytics

>     metadata:

Reviewer: Adding @ahmelsayed for input on the ForEnv renaming we are doing

Author: @tomkerkhove , I have added support for 'FromEnv' in last commit for scaler.
CC: @ahmelsayed

Reviewer: Thanks!

Reviewer: Can we update the docs here to list those new properties as well please?

Author: @tomkerkhove , I have added description of 'FromEnv' in last commit.

      tenantID: "72f988bf-86f1-41af-91ab-2d7cd011db47"
      clientID: "04b4ca0a-82b1-4c0a-bbbb-7946442e805b"
      clientSecret: "vU6UtUXls6RNXxv~l6NRi1V8J1fnk5Q-ce"
      workspaceID: "81963c40-af2e-47cd-8e72-3002e08aa2af"

>       query: "let AppName = \"web\";\r\nlet ClusterName = \"demo-cluster\";\r\nlet AvgDuration = ago(10m);\r\nlet ThresholdCoefficient = 0.8;\r\nPerf\r\n| where InstanceName contains AppName\r\n| where InstanceName contains ClusterName\r\n| where CounterName == \"cpuUsageNanoCores\"\r\n| where TimeGenerated > AvgDuration\r\n| extend AppName = substring(InstanceName, indexof((InstanceName), \"/\", 0, -1, 10) + 1)\r\n| summarize MetricValue=round(avg(CounterValue)) by CounterName, AppName\r\n| join (Perf \r\n        | where InstanceName contains AppName\r\n        | where InstanceName contains ClusterName\r\n        | where CounterName == \"cpuLimitNanoCores\"\r\n        | where TimeGenerated > AvgDuration\r\n        | extend AppName = substring(InstanceName, indexof((InstanceName), \"/\", 0, -1, 10) + 1)\r\n        | summarize arg_max(TimeGenerated, *) by AppName, CounterName\r\n        | project Limit = CounterValue, TimeGenerated, CounterPath, AppName)\r\n        on AppName\r\n| project MetricValue, Threshold = Limit * ThresholdCoefficient"

Reviewer: I'm wondering - Do we use in-line queries, use config map or mount them on the scaled workload?

Reviewer: Thoughts @zroubalik @ahmelsayed ?

Author: @tomkerkhove , I used Prometheus Scaler as reference: https://keda.sh/docs/1.5/scalers/prometheus/

Reviewer: Makes sense but in this case, they are a bit longer. It's ok for me but just want to verify with the others.

Reviewer 2: You can define a multi-line string in yaml using 

```yaml
      ...
      query: |
        let AppName = "web";
        let ClusterName = "demo-cluster";
        let AvgDuration = ago(10m);
        let ThresholdCoefficient = 0.8;
        Perf
        | where InstanceName contains AppName
        | where InstanceName contains ClusterName
        | where CounterName == "cpuUsageNanoCores"
        | where TimeGenerated > AvgDuration
        | extend AppName = substring(InstanceName, indexof((InstanceName), "/", 0, -1, 10) + 1)
        | summarize MetricValue=round(avg(CounterValue)) by CounterName, AppName
        | join (Perf 
                | where InstanceName contains AppName
                | where InstanceName contains ClusterName
                | where CounterName == "cpuLimitNanoCores"
                | where TimeGenerated > AvgDuration
                | extend AppName = substring(InstanceName, indexof((InstanceName), "/", 0, -1, 10) + 1)
                | summarize arg_max(TimeGenerated, *) by AppName, CounterName
                | project Limit = CounterValue, TimeGenerated, CounterPath, AppName)
                on AppName
        | project MetricValue, Threshold = Limit * ThresholdCoefficient
```

which might make it look a bit more readable.

Reviewer: Love the suggestion!

      threshold: "1900000000"
```

**Parameter list:**

> - `tenantID`: Azure Active Directory TenantId. Follow [this](https://docs.microsoft.com/en-us/cli/azure/account?view=azure-cli-latest#az-account-show) link to retrieve your TenantId.

Reviewer: Reads a bit more fluent
```suggestion
- `tenantID`: Azure Active Directory tenant id. Follow [this](https://docs.microsoft.com/en-us/cli/azure/account?view=azure-cli-latest#az-account-show) link to retrieve your tenant id.

```

> - `clientID`: appId from your Service Principal. Follow [this](https://docs.microsoft.com/en-us/cli/azure/ad/sp?view=azure-cli-latest) link to create your Service Principal.

Reviewer: Reads a bit more fluent
```suggestion
- `clientID`: Application id from your Azure AD Application/service principal. Follow [this](https://docs.microsoft.com/en-us/cli/azure/ad/sp?view=azure-cli-latest) link to create your service principal.

```

> - `clientSecret`: password from your Service Principal.

Reviewer: ```suggestion
- `clientSecret`: Password from your Azure AD Application/service principal.

```

> - `workspaceID`: Your Log Analytics WorkspaceId. Follow [this](https://docs.microsoft.com/en-us/cli/azure/monitor/log-analytics/workspace?view=azure-cli-latest#az-monitor-log-analytics-workspace-list) link to get your Log Analytics WorkspaceId.

Reviewer: ```suggestion
- `workspaceID`: Your Log Analytics workspace id. Follow [this](https://docs.microsoft.com/en-us/cli/azure/monitor/log-analytics/workspace?view=azure-cli-latest#az-monitor-log-analytics-workspace-list) link to get your Log Analytics workspace id.

```

- `query`: Log Analytics [kusto](https://docs.microsoft.com/en-us/azure/azure-monitor/log-query/get-started-queries) query, JSON escaped. You can use [this](https://www.freeformatter.com/json-escape.html) tool to convert your query from Log Analytics query editor to JSON escaped string, and then review YAML specific escapes.
- `threshold`: An int64 value will be used as a threshold to calculate # of pods for scale target.

> **Threshold and query results:**

Reviewer: Maybe move this to a dedicated section?
```suggestion
### Query Guidance
```

Or:
```suggestion
### Query Requirements
```

It is important to design your query to return 1 table with 1 row. A good practice is to add "| limit 1" at the end of your query.

> Scaler will take value from:

Reviewer: Required for rendered docs:
```suggestion

Scaler will take value from:

```

```
let AppName = "web";
let ClusterName = "demo-cluster";
let AvgDuration = ago(10m);
let ThresholdCoefficient = 0.8;
Perf
| where InstanceName contains AppName
| where InstanceName contains ClusterName
| where CounterName == "cpuUsageNanoCores"
| where TimeGenerated > AvgDuration
| extend AppName = substring(InstanceName, indexof((InstanceName), "/", 0, -1, 10) + 1)
| summarize MetricValue=round(avg(CounterValue)) by CounterName, AppName
| join (Perf 
        | where InstanceName contains AppName
        | where InstanceName contains ClusterName
        | where CounterName == "cpuLimitNanoCores"
        | where TimeGenerated > AvgDuration
        | extend AppName = substring(InstanceName, indexof((InstanceName), "/", 0, -1, 10) + 1)
        | summarize arg_max(TimeGenerated, *) by AppName, CounterName
        | project Limit = CounterValue, TimeGenerated, CounterPath, AppName)
        on AppName
| project MetricValue, Threshold = Limit * ThresholdCoefficient
```

You can convert your query in JSON escaped string using [this](https://www.freeformatter.com/json-escape.html) tool and check yaml specific escapes.

> **Scaler limitations (READ CAREFULLY):**

Reviewer: Maybe move this to a `###` header after the `Authentication Parameters` section?

```suggestion
### Scaler Limitations
```

- As it was mentioned before, you can define a threshold using query (2d cell of query result will be interpret as threshold). Be aware! Threshold from query result will be set only once, during scaler creation. So, if your query will return different threshold values during runtime, they will not be propagated to HPA target.
  
### Authentication Parameters

 You can use `TriggerAuthentication` CRD to configure the `tenantID`, `clientID`, `clientSecret` and `workspaceID` to connect to Log Analytics API.

> **Service Principal based authentication:**

Reviewer: Can you align with other scalers here please?

https://keda.sh/docs/1.5/scalers/azure-monitor/#authentication-parameters

For example, provide more fluent sentences (example: `appId` -> `ID of Azure AD Application`)

Author: Done in last commit

Reviewer: Thanks!

- `tenantID` required: Azure Active Directory TenantId. Follow [this](https://docs.microsoft.com/en-us/cli/azure/account?view=azure-cli-latest#az-account-show) link to retrieve your TenantId.
- `clientID` required: appId from your Service Principal. Follow [this](https://docs.microsoft.com/en-us/cli/azure/ad/sp?view=azure-cli-latest) link to create your Service Principal.
- `clientSecret` required: password from your Service Principal.
- `workspaceID` required: Your Log Analytics WorkspaceId. Follow [this](https://docs.microsoft.com/en-us/cli/azure/monitor/log-analytics/workspace?view=azure-cli-latest#az-monitor-log-analytics-workspace-list) link to get your Log Analytics WorkspaceId.

> ### Example

Reviewer: In this case, it might be useful to include a screenshot of Kusto query result in Azure portal, but maybe that's a bit too far?

Author: Added screenshot in last commit, but in "### Query Guidance" section.

Reviewer: Thanks!

- `query`: Log Analytics [kusto](https://docs.microsoft.com/en-us/azure/azure-monitor/log-query/get-started-queries) query, JSON escaped. You can use [this](https://www.freeformatter.com/json-escape.html) tool to convert your query from Log Analytics query editor to JSON escaped string, and then review YAML specific escapes.
- `threshold`: An int64 value will be used as a threshold to calculate # of pods for scale target.

### Query Guidance

It is important to design your query to return 1 table with 1 row. A good practice is to add "| limit 1" at the end of your query.

Scaler will take value from:

- 1st cell as Metrics Value.
- 2d cell as Threshold (optional).

You can define threshold in trigger metadata, it will be used if your query results only 1 cell, that will be interpreted as metric value. Be aware, even if you have defined threshold in metadata, it can be overwritten by your query.

A data types of your query result should be: real, int or long. Other data types are not supported. Later, during runtime, your data will be converted to int64.

Be careful with setting up "pollingInterval" and long running queries. Test your query before.

> Example query to get MetricValue and Threshold based on CPU usage and Limits, defined for the pod.

Reviewer: ```suggestion
Example query to get `MetricValue` and `Threshold` based on CPU usage and limits, defined for the pod.

```
